From 868f1231bfc29295c032646fc22e79273e61e62e Mon Sep 17 00:00:00 2001
From: Daniel Ammann <daniel.ammann@bytesatwork.ch>
Date: Tue, 6 Sep 2022 08:29:28 +0200
Subject: [PATCH] Update github action configuration for the tool that cannot
 be named

Update checkpatch.yml to 2.0.0: Check each patch file in a separate job.

Signed-off-by: Daniel Ammann <daniel.ammann@bytesatwork.ch>
---
 .github/workflows/checkpatch.yml | 97 ++++++++++++++++++++++++++++++--
 1 file changed, 92 insertions(+), 5 deletions(-)

diff --git a/.github/workflows/checkpatch.yml b/.github/workflows/checkpatch.yml
index 8d8692778a28..ff8136931937 100644
--- a/.github/workflows/checkpatch.yml
+++ b/.github/workflows/checkpatch.yml
@@ -4,29 +4,116 @@
 # Changelog
 #
 # [1.0.0] - 2022-03-31: Initial version
+# [1.0.1] - 2022-07-15: Install required Python packages 'ply' and 'GitPython'
+# [1.1.0] - 2022-07-18: Extend to handle Zephyr project repos as well
+# [1.2.0] - 2022-09-02: Extend to handle Zephyr module repos as well
+# [2.0.0] - 2022-09-02: Check each patch file in a separate job
 
 
 # This Github action checks a pull request using checkpatch
 name: checkpatch
 on: pull_request
 jobs:
-  checkpatch:
-    name: Checkpatch
+  setup-checkpatch:
+    name: Setup Checkpatch
     runs-on: ubuntu-latest
+    env:
+      ZEPHYR_DEFAULT_VERSION: v3.1.0
+    outputs:
+      matrix: ${{ steps.set-matrix.outputs.patches }}
 
     steps:
       - name: Checkout repository
-        uses: actions/checkout@v2
+        uses: actions/checkout@v3
         with:
           fetch-depth: 100
           ref: ${{ github.event.pull_request.head.sha }}
 
       - name: Install dependencies
-        run: sudo apt-get update && sudo apt-get install -y perl
+        run: |
+          sudo apt-get update && sudo apt-get install -y jq
+
+      - name: Checkout Zephyr project repository
+        # if the repository contains a Zephyr project (if west.yml exists)
+        # or a Zephyr module (if zephyr folder exists), the Zephyr project
+        # repository must be cloned in order to get the checkpatch script
+        run: |
+          if [[ -f "west.yml" ]]
+          then
+            ZEPHYR_VERSION=$(grep -A 3 "name: zephyr$" west.yml | grep "revision:" | grep -Eo "v[0-9]+\.[0-9]+\.[0-9]+") || true
+            if [[ -z "${ZEPHYR_VERSION}" ]]
+            then
+              ZEPHYR_VERSION=${ZEPHYR_DEFAULT_VERSION}
+            fi
+          else
+            ZEPHYR_VERSION=${ZEPHYR_DEFAULT_VERSION}
+          fi
+          git clone --depth 1 --branch ${ZEPHYR_VERSION} https://github.com/zephyrproject-rtos/zephyr.git zephyr-rtos
+        if: hashFiles('west.yml') != '' || hashFiles('zephyr/*') != ''
 
       - name: Create patch file(s)
         # create patch file(s) of new commits (between base and head)
         run: git format-patch ${{ github.event.pull_request.base.sha }}..${{ github.event.pull_request.head.sha }} || exit 1
 
+      - name: Store files to cache
+        # cache all files required for checkpatch to make them available for other jobs
+        uses: actions/cache@v3
+        with:
+          path: |
+            ${{ github.workspace }}/scripts/*
+            ${{ github.workspace }}/zephyr-rtos/scripts/*
+            ${{ github.workspace }}/.checkpatch.conf
+            ${{ github.workspace }}/zephyr-rtos/.checkpatch.conf
+            ${{ github.workspace }}/west.yml
+            ${{ github.workspace }}/zephyr/*
+            ${{ github.workspace }}/*.patch
+          key: checkpatch-files-${{ hashFiles('*.patch') }}
+
+      - name: Set list of patch filenames as output
+        id: set-matrix
+        run: |
+          echo "::set-output name=patches::$(ls *.patch | jq -R -s -c 'split("\n")[:-1]')"
+
+
+  run-checkpatch:
+    name: Run Checkpatch
+    runs-on: ubuntu-latest
+    needs: setup-checkpatch
+    strategy:
+      fail-fast: false
+      matrix:
+        patch: ${{ fromJSON(needs.setup-checkpatch.outputs.matrix) }}
+
+    steps:
+      - name: Get files from cache
+        # restore all cached files required for checkpatch
+        uses: actions/cache@v3
+        with:
+          path: |
+            ${{ github.workspace }}/scripts/*
+            ${{ github.workspace }}/zephyr-rtos/scripts/*
+            ${{ github.workspace }}/.checkpatch.conf
+            ${{ github.workspace }}/zephyr-rtos/.checkpatch.conf
+            ${{ github.workspace }}/west.yml
+            ${{ github.workspace }}/zephyr/*
+            ${{ github.workspace }}/*.patch
+          key: checkpatch-files-${{ hashFiles('*.patch') }}
+
+      - name: Install dependencies
+        run: |
+          sudo apt-get update && sudo apt-get install -y perl python3
+          python3 -m pip install ply
+          python3 -m pip install GitPython
+
       - name: Run checkpatch
-        run: ./scripts/checkpatch.pl *.patch
+        # if the repository contains a Zephyr project (if west.yml exists)
+        # or a Zephyr module (if zephyr folder exists), the checkpatch script
+        # is located in the Zephyr folder
+        run: |
+          if [[ -f "west.yml" || -d "zephyr" ]]
+          then
+            cd zephyr-rtos
+            ./scripts/checkpatch.pl --no-tree ../${{ matrix.patch }}
+          else
+            ./scripts/checkpatch.pl --no-tree ${{ matrix.patch }}
+          fi
-- 
2.30.2

